src/chat/KeywordResponses.cpp

```diff
@@ -0,0 +1,48 @@
+// akey.cpp
+
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+
+#include "KeywordResponses.hpp"
+
+using namespace chat;
+
+namespace {
+
+int random(int x) { return rand() % x; }
+
+}
+
+const char *KeywordResponses::get_random_response() {
+    const char *defaultmsg = "Mmm, sounds interesting...";
+    if (total_weight_ == 0)
+        return defaultmsg;
+    int rnd_num = random(total_weight_), total = 0;
+    for (int i = 0; i < num_responses_; i++) {
+        if ((total += (responses_[i].weight_)) > rnd_num) {
+            return responses_[i].message_;
+        }
+    }
+    return defaultmsg;
+}
+
+int KeywordResponses::add_keywords(char *keyword) { return (keywords_ = strdup(keyword)) != nullptr; }
+
```

Comment:
I'd like to get rid of this nasty alloc business.  Thinking vector.

Reply:
right, and std::string